test/extensions/watchdog/profile_action/profile_action_test.cc

```diff
@@ -0,0 +1,328 @@
+    action_ = std::make_unique<ProfileAction>(config, context_);
+  }
+
+  void waitForOutstandingNotify() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+    mutex_.Await(absl::Condition(
+        +[](int* outstanding_notifies) -> bool { return *outstanding_notifies > 0; },
+        &outstanding_notifies_));
+    outstanding_notifies_ -= 1;
+  }
+
+  std::unique_ptr<Event::TestTimeSystem> time_system_;
+  Api::ApiPtr api_;
+  Event::DispatcherPtr dispatcher_;
+  Server::Configuration::GuardDogActionFactoryContext context_;
+  std::unique_ptr<Server::Configuration::GuardDogAction> action_;
+  Thread::ThreadPtr thread_;
+  // Path for the test case to dump any profiles to.
+  const std::string test_path_;
+  // Used to synchronize with the dispatch thread
+  absl::Mutex mutex_;
+  int outstanding_notifies_ ABSL_GUARDED_BY(mutex_) = 0;
+};
+
+INSTANTIATE_TEST_SUITE_P(TimeSystemType, ProfileActionTest,
+                         testing::ValuesIn({TimeSystemType::Real, TimeSystemType::Simulated}));
+
+TEST_P(ProfileActionTest, CanDoMultipleProfiles) {
+  // Create configuration.
+  envoy::extensions::watchdog::profile_action::v3alpha::ProfileActionConfig config;
+  TestUtility::loadFromJson(absl::Substitute(R"EOF({
+          "profile_duration": "1s",
+          "profile_path": "$0",
+        }
+      )EOF",
+                                             test_path_),
+                            config);
+  setupAction(config);
+
+  // Create vector of relevant threads
+  const auto now = api_->timeSource().monotonicTime();
+  std::vector<std::pair<Thread::ThreadId, MonotonicTime>> tid_ltt_pairs = {
+      {Thread::ThreadId(10), now}};
+
+  // Check that we can do at least a single profile
+  dispatcher_->post([&tid_ltt_pairs, &now, this]() -> void {
+    action_->run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::MISS, tid_ltt_pairs, now);
+    absl::MutexLock lock(&mutex_);
+    outstanding_notifies_ += 1;
+  });
+
+  absl::MutexLock lock(&mutex_);
+  waitForOutstandingNotify();
+  time_system_->advanceTimeWait(std::chrono::seconds(2));
+
+#ifdef PROFILER_AVAILABLE
+  EXPECT_EQ(countNumberOfProfileInPath(test_path_), 1);
+#else
+  // Profiler won't run in this case, so there should be no files generated.
+  EXPECT_EQ(countNumberOfProfileInPath(test_path_), 0);
+#endif
+
+  // Check we can do multiple profiles
+  dispatcher_->post([&tid_ltt_pairs, &now, this]() -> void {
+    action_->run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::MISS, tid_ltt_pairs, now);
+    absl::MutexLock lock(&mutex_);
+    outstanding_notifies_ += 1;
+  });
+
+  waitForOutstandingNotify();
+  time_system_->advanceTimeWait(std::chrono::seconds(2));
+
+  dispatcher_->exit();
+  thread_->join();
+
+#ifdef PROFILER_AVAILABLE
+  EXPECT_EQ(countNumberOfProfileInPath(test_path_), 2);
+#else
+  // Profiler won't run in this case, so there should be no files generated.
+  EXPECT_EQ(countNumberOfProfileInPath(test_path_), 0);
+#endif
+}
```

Comment:
This seems like a separate behavior than the single-profile case. Could/should this be a separate test? If not, should the EXPECT_EQ above be asserts?

Reply:
Will split out the single profile case from this.